mitmproxy/proxy/layers/dns.py

```diff
@@ -37,16 +41,30 @@ class DnsErrorHook(commands.StartHook):
     flow: dns.DNSFlow
 
 
+def pack_message(
+    message: dns.Message, transport_protocol: Literal["tcp", "udp"]
+) -> bytes:
+    packed = message.packed
+    if transport_protocol == "tcp":
+        return struct.pack("!H", len(packed)) + packed
+    else:
+        return packed
+
+
 class DNSLayer(layer.Layer):
     """
     Layer that handles resolving DNS queries.
     """
 
     flows: dict[int, dns.DNSFlow]
+    req_buf: bytearray
+    resp_buf: bytearray
```

Comment:
FWIW, bytearray is great here - things are small anyhow, and mutability is useful for this.